packages/@uppy/thumbnail-generator/src/index.js

```diff
@@ -303,12 +312,25 @@ module.exports = class ThumbnailGenerator extends Plugin {
 
   onFileAdded = (file) => {
     if (!file.preview && isPreviewSupported(file.type) && !file.isRemote) {
-      this.addToQueue(file)
+      this.addToQueue(file.id)
     }
   }
 
+  /**
+   * Cancel a lazy request for a thumbnail if the thumbnail has not yet been generated.
+   */
+  onCancelRequest = (file) => {
+    const index = this.queue.indexOf(file.id)
+    if (index !== -1) {
+      this.queue.splice(index, 1)
+    }
+  }
+
+  /**
+   * Clean up the thumbnail for a file. Cancel lazy requests and free the thumbnail URL.
+   */
   onFileRemoved = (file) => {
-    const index = this.queue.indexOf(file)
+    const index = this.queue.indexOf(file.id)
```

Comment:
Storing file IDs in the `this.queue` array now because of this: previously items would not be removed from the queue because the file object could be different (eg. metadata was changed or upload progress has happened), and then `indexOf()` would not find it.